Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp

```diff
@@ -103,142 +76,76 @@ namespace AZ
-                // the parser to force parsing remaining arguments as positional
-                if (currentArg == PositionalArgSeparator)
-                {
-                    argumentParserState.m_parseRemainAsPositional = true;
-                    argumentParserState.m_currentOption.clear();
-                    return;
-                }
-
-                // If the `--` argument has been parsed, treat all remaining arguments positional
-                if (m_commandLineOptionPrefix.contains(currentArg.front()))
-                {
-                    // its possible that its a key-value-pair like -blah=whatever
-                    // we support this too, for compatibility.
-
-                    currentArg = currentArg.substr(1);
-                    if (currentArg[0] == '-') // for -- extra
-                    {
-                        currentArg = currentArg.substr(1);
-                    }
-
-                    AZStd::size_t foundPos = AZ::StringFunc::Find(currentArg, "=");
-                    if (foundPos != AZStd::string::npos)
-                    {
-                        AZStd::string_view argumentView{ currentArg };
-                        AZStd::string_view option = AZ::StringFunc::StripEnds(argumentView.substr(0, foundPos));
-                        AZStd::string_view value = AZ::StringFunc::StripEnds(argumentView.substr(foundPos + 1));
-                        ParseOptionArgument(ToLower(option), value, nullptr);
-                        argumentParserState.m_currentOption.clear();
-                    }
-                    else
-                    {
-                        // its in this format -switchName switchvalue
-                        // (no equals)
-                        argumentParserState.m_currentOption = ToLower(currentArg);
-                        m_allValues.push_back({ argumentParserState.m_currentOption, "" });
-                    }
-                    return;
-                }
-            }
-
-            if (argumentParserState.m_currentOption.empty())
-            {
-                // Parse positional argument
-                m_allValues.push_back({ "", UnquoteArgument(currentArg) });
-            }
-            else
-            {
-                // Finish parsing of values for option argument
-                ParseOptionArgument(argumentParserState.m_currentOption, currentArg, &m_allValues.back());
-                argumentParserState.m_currentOption.clear();
+                m_allValues.push_back({ ToLower(newOption), optionValue });
             }
         }
     }
 
     void CommandLine::Parse(int argc, char** argv)
     {
         m_allValues.clear();
-
-        // Stores the state of arguments being parsed
-        // Allows the AddArgument function to update the state
-        ArgumentParserState parseState;
-
-        // Start on 1 because 0 is the executable name
-        for (int i = 1; i < argc; ++i)
+        if (argc == 0)
```

Comment:
When would this case be hit? Won't argc always be at least 1, with argv[0] being the executable name?
I'm not opposed to having the early out here anyways, it just seems unnecessary.

Reply:
Well the code doesn't assume that the command line parameters are being forwarded from the launch parameters of the application through `int main()`.

The user is allowed to construct an in-memory set of command line parameters and parse them.
That is something the UnitTest also checks for.

Reply:
Heh, TIL that the C standard defines that argc==0 is allowed. Apparently it can happen in some hosted environments. But _if_ argc > 0, then argv[0] is expected to contain the program name. IMO, anything calling a method with argc/argv parameters should follow this same definition, whether constructed in-memory, or unit tests, or passed from the command line. But regardless, the early out makes sense to me now, ty!

For reference:
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf 
```
5.1.2.2.1 Program startup
1 The function called at program startup is named main. The implementation declares no
prototype for this function. It shall be defined with a return type of int and with no
parameters:
int main(void) { /* ... */ }
or with two parameters (referred to here as argc and argv, though any names may be
used, as they are local to the function in which they are declared):
int main(int argc, char *argv[]) { /* ... */ }
or equivalent;10) or in some other implementation-defined manner.
2 If they are declared, the parameters to the main function shall obey the following
constraints:
— The value of argc shall be nonnegative.
— argv[argc] shall be a null pointer.
— If the value of argc is greater than zero, the array members argv[0] through
argv[argc-1] inclusive shall contain pointers to strings, which are given
implementation-defined values by the host environment prior to program startup. The
intent is to supply to the program information determined prior to program startup
from elsewhere in the hosted environment. If the host environment is not capable of
supplying strings with letters in both uppercase and lowercase, the implementation
shall ensure that the strings are received in lowercase.
— If the value of argc is greater than zero, the string pointed to by argv[0]
represents the program name; argv[0][0] shall be the null character if the
program name is not available from the host environment. If the value of argc is
greater than one, the strings pointed to by argv[1] through argv[argc-1]
represent the program parameters.
— The parameters argc and argv and the strings pointed to by the argv array shall
be modifiable by the program, and retain their last-stored values between program
startup and program termination.
```